### ISSUE TITLE
Split out Unix-specific code to support MirageOS better

### ISSUE THREAD

**Issue by cfcs** [rlepigre/ocaml-imagelib#17]
This is a proposal to split out the Unix-specific function calls to a `imagelib.unix` package, to enable use (well, linking) on MirageOS.

I have not had too many issues since I mainly use MirageOS on Xen, which doesn't seem to do *eager relocations,* but @hannesm mentioned that this is a problem on some of the other backends.

In any case the problem is that functions like `Sys.remove`, `Sys.command`, `Unix.create_process`, [...] are not available in MirageOS, and calling them causes the application to crash.

**Comment by rlepigre** [comment]
OK, we need to agree on what everyone wants before we start messing around with the interface of the library. In particular, we need to define a general interface for the implementation of image formats, independent from any particular input/output directives. Then, we could probably build several possible back-ends on top of that, for handling higher-level function (e.g., reading an image file from an arbitrary format).

What does everyone think?

**Comment by cfcs** [comment]
@rlepigre I think that sounds perfect :)

(For future reference):
Part of the motivation proposal was also discussed here https://github.com/rlepigre/ocaml-imagelib/pull/15#issuecomment-401420541

@yomimono Thanks for bringing up `js_of_ocaml`, I think it's a super good point that both `js_of_ocaml` and MirageOS share that concern that they do not provide the Unix API.
My humble attempt to implement reading from chunks vs files was merged here, is this similar to your patch? https://github.com/rlepigre/ocaml-imagelib/pull/8

**Comment by cfcs** [comment]
I'm working on the proposed split here, incorporating the Dune transition:
https://github.com/cfcs/ocaml-imagelib/tree/mirage_compat

I've still yet to try to compile this on mirage, so I'm not sure if it works, or if there are more `Unix` dependencies hiding in there, and I created a bit of a mess since I only noticed the `OUT_CHANNEL` stuff after implementing a `chunk_writer`. I'll try to consolidate that with the `OUT_CHANNEL` signature (maybe we should also have a `IN_CHANNEL` signature?).

The branch seems to compile, but it `writefile` on Unix generates half-broke files (only the first part of the image shows).
